src/world/conquest_system.cpp

```diff
@@ -23,36 +23,50 @@ along with this program.  If not, see http://www.gnu.org/licenses/
 
 #include "message_server.h"
 
+/**
+ * ConquestSystem both handles messages from map servers and
+ * updates the database with the latest conquest data periodically.
+ *
+ * This class is guided by the following pattern:
+ * - All public methods that may modify the database are enqueued in the task system.
+ * - The task system processes tasks in its own thread
+ * - Private methods are not guarded via the task system, but are only called from
+ *   public methods, so we can assume that they are always called from the task system.
+ */
 ConquestSystem::ConquestSystem()
-: sql(std::make_unique<SqlConnection>())
 {
+    submit([this]()
+           { sql = std::make_unique<SqlConnection>(); });
```

Comment:
I don't think task_system is going to work for your use case here if you have to submit a task to configure it - I think it's too simple of a building block to achieve what you want.

Reply:
What _might_ work here, is if you make IMessageHandler also inherit from task_system, giving you access to the guts and the construction of everything. This would allow you to lazily set up the sql handle upon submitting the first task, or on construction, or whatever.

Reply:
and just because this was the review point that made me remember the term I've been trying to remember for _WEEKS_: https://en.wikipedia.org/wiki/Convention_over_configuration

Reply:
So been thinking about this. I wouldn't need to do this if it wasn't for the warning that pops up when checking if sql query is executed in the same thread it's instantiated. 

Because all queries are going to happen in the same thread (the task system thread), it's okay that we instantiate sql in a different thread. Ideally, the check would only happen if two queries are called from different threads. Would you be okay with that change?

Reply:
Ah I do see that sqlconnection fires a query on instantiation, which makes this change harder

Reply:
Absolutely not. I don't play up my 'trap card' to be cute or funny, I do it because it's a legitimate problem that was very annoying to diagnose and fix. https://github.com/LandSandBoat/server/blob/b61c1498f7c639a06bcb1822eaf97d8c4368b1d1/src/common/sql.h#L30

Reply:
The warning triggers in the exact scenario that results in UB, most often seen as the sql handle getting wiped out - so don't ignore it